### ISSUE TITLE
ClassCastException in "isArray" (JsonAssertList)

### ISSUE THREAD

**Issue by glhez** [lukas-krecan/JsonUnit#474]
I am using AssertJ with 3.22.0 (latest at date) and json-unit 2.31.0

The following code was working before:

```java 
    @Test
    void test_cce_error() {
       // import static net.javacrumbs.jsonunit.assertj.JsonAssertions.assertThatJson;
        String json = "[ [ \"a\", \"b\", \"c\", \"d\","
                + " \"e\", \"f\", \"g\", \"h\", \"i\" ] ]";
        assertThatJson(json).inPath("$.[0]")
                            .isArray()
                            .elements(3 /* d */, 4 /* e */, 5 /* f */ , 6 /* g */, 7 /* h */, 8 /* i */ )
                            .containsExactly("d", "e", "f", "g", "h", "i");
    }
```

However, this no longer work due to change in assertj:

```
java.lang.ClassCastException: org.assertj.core.api.FactoryBasedNavigableListAssert cannot be cast to net.javacrumbs.jsonunit.assertj.JsonListAssert
```

I tried to pinpoint the issue: the `elements` is creating a new list out of the current. By default (= case of JsonAssertList) it uses this implements:

``` // in class org.assertj.core.api.FactoryBasedNavigableListAssert<SELF, ACTUAL, ELEMENT, ELEMENT_ASSERT>
  @SuppressWarnings("unchecked")
  @Override
  protected SELF newAbstractIterableAssert(Iterable<? extends ELEMENT> iterable) {
    checkArgument(iterable instanceof List, "Expecting %s to be a List", iterable);
    return (SELF) new FactoryBasedNavigableListAssert<>((List<? extends ELEMENT>) iterable,
                                                        FactoryBasedNavigableListAssert.class,
                                                        assertFactory);
  }
 ```

I think that JsonAssertList should implement it and probably that assertj-core should make the class abstract and force an implementation of said method.

If possible, I'll try to attach a PR with test.

**Comment by lukas-krecan** [comment]
Thanks for feedback, I will look at it in next few days. If you can provide a PR, I would be grateful.

**Comment by glhez** [comment]
I'm not sure _all_ problems but mine are fixed - I for example don't think the JSON path is still valid here. 

I also fixed a problem on [JsonMapAssert](https://github.com/glhez/JsonUnit/commit/1d83fb966ef705099134f04ab25865183206bb83#diff-b9dab470e169894d74606060b716eb03d90e11c5019e40fd8aab880487576334) where method signature are not same as parent class. This might work with Javac but fails with ECJ.

**Comment by glhez** [comment]
FYI, this also affect isArray().first() because it also use toAssert.

**Comment by lukas-krecan** [comment]
Hi, I have fixed the bug. I have a question for you. It seems you want to mix Json assertions and AssertJ methods. Would it be useful for you to have a method `toAssertJ()` that would let you move from Json comparison to pure AssertJ methods? It's not clear from the example with InstanceOfAssertFactories if you want to comapre the values as Json or as string, `toAssertJ()`  method would make it explicit.

**Comment by glhez** [comment]
> Hi, I have fixed the bug.

Thanks :)

>
> I have a question for you. It seems you want to mix Json assertions and AssertJ methods. Would it be useful for you to have a method toAssertJ() that would let you move from Json comparison to pure AssertJ methods?

I don't think you need to add a "toAssertJ()" method. 

And in all honesty, it is not like I decided to mix the two, more that I decided to use the AssertJ variant of JsonUnit and that I've found the `elements(int...)` and `element(int, InstanceOfAssertFactory)` I've used a lot of time with AssertJ when dealing with iterable.

Only it failed after a dependency update :)

I'm acustomized to using `InstanceOfAssertFactory` with List/Map assertion from AssertJ which don't (or rather can't) provide a customized element assert such as `JsonAssert` and I will use the `isString()`.

The problem in the case of JsonUnit lies in the fact that JsonUnit APIs are interwined with the AssertJ API and subject to usage/implementation issues: some method are (sadly) not abstract in AssertJ, and don't make sense as is.

The one I reported here directly casted a type "SELF" (= `JsonListAssert`) to `FactoryBasedNavigableListAssert` which resulted in this `ClassCastException` (this probably is no longer possible to make them abstract in AssertJ without breaking changes :)).

> It's not clear from the example with InstanceOfAssertFactories if you want to compare the values as Json or as string, toAssertJ() method would make it explicit.

As said, you don't need to.

I've wrote some example to (perhaps) complete the README and my tests:

```java
String json = "[ [ \"a\", 7 ] ]";
// WORK
assertThatJson(json).inPath("$.[0]")
                    .isArray()
                      .elements(0)
                        .isString()
                          .startsWith("a");

// WORK
assertThatJson(json).inPath("$.[0]")
                    .isArray()
                      .elements(0, 1)
                        .containsExactly("A", 7);

String json2 = "[ [ \"a\", 31, [ \"c\" ], \"d\", \"e\", \"f\", \"g\", \"h\", \"i\" ] ]";
// WORK
assertThatJson(json2).inPath("$.[0]")
                    .isArray()
                      .elements(0, InstanceOfAssertFactories.STRING)
                        .startsWith("a");
// FAIL Integer != BigDecimal
assertThatJson(json2).inPath("$.[0]")
                     .isArray()
                     .element(1, InstanceOfAssertFactories.INTEGER)
                     .isEqualTo(31);
// FAIL Object[] != JsonList
assertThatJson(json2).inPath("$.[0]")
                     .isArray()
                     .element(2, InstanceOfAssertFactories.ARRAY)
                     .containsOnly("c");
// WORK
Object[] values = { new Object[] { "c" } };
assertThatJson(json).inPath("$.[0]")
                    .isArray()
                    .elements(2)
                    .containsExactly(values);

// FAILS
// ["c"]
// - element index: 1
// - error:
// expected: "["c"] (Object[]@34f5090e)"
//  but was: "["c"] (JsonList@693fe6c9)"
Object[] values = { "c" };
assertThatJson(json).inPath("$.[0]")
                    .isArray()
                    .elements(3, 2)
                    .satisfiesExactly( d -> assertThat(d).isEqualTo("d")
                                     , c -> assertThat(c).isEqualTo(values));

// WORKS
assertThatJson(json).inPath("$.[0]")
                    .isArray()
                    .elements(3, 2)
                    .satisfiesExactly( d -> assertThat(d).isEqualTo("d")
                                     , c -> assertThatJson(c).isEqualTo(values));

```

After writing them, here is what I can say about the AssertJ API I used:

The [`element(int, InstanceOfAssertFactory)`][JAVADOC-003] (and the first/last family) will probably fails to work as intended when the type is NOT a string (or I suppose, boolean).

[asInstanceOf][JAVADOC-004] can be overriden in `JsonAssert` as below  to avoid misuse of the API especially when the method is provided by AssertJ `AbstractAssert` and for which you can't do anything about it.

```java
@Override
public <ASSERT extends AbstractAssert<?, ?>> ASSERT asInstanceOf(InstanceOfAssertFactory<?, ASSERT> instanceOfAssertFactory) {
    throw failure("Please use isString(), isNumber(), isBoolean(), isNull(), isUri(), isArray() or isObject().%n"
            + "This method will most likely not provide the result you expect it to.");
}
```

This would give a good insight of an API usage issue (a bit like what Mockito does when the mock is not used/etc).

The `satisfiesExactly()` is similar to `JsonAssertions.assertThatJson(Object, JsonAssertionCallback...)`, but the problem is that you can't use either only `assertThat`, either only `assertThatJson`: the first expression will fail with a `Can not parse actual value: 'd'`.

The [XXXForProxy][JAVADOC-005] method could probably be overriden in `JsonListAssert` (and probably `JsonAssertMap`) to perform a conversion of ELEMENT to a proper non-json type before delegating to the consumers.

But the `assertThatJson(json2).and(...)` read better in this case:

```java
String json2 = "[ [ \"a\", 31, [ \"c\" ], \"d\", \"e\", \"f\", \"g\", \"h\", \"i\" ] ]";
assertThatJson(json2).and(
  a -> a.node("$.[0][3]").isString().isEqualTo("d"),
  a -> a.node("$.[0][2]").isArray().containsOnly("c")
);
```

So the, `satisfiesXXX` methods could be probably be overriden/deprecated with a javadoc explaining use case.


[JAVADOC-003]: https://www.javadoc.io/static/org.assertj/assertj-core/3.22.0/org/assertj/core/api/AbstractIterableAssert.html#element(int,org.assertj.core.api.InstanceOfAssertFactory)
[JAVADOC-004]: https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/AbstractAssert.html#asInstanceOf(org.assertj.core.api.InstanceOfAssertFactory)
[JAVADOC-005]: https://www.javadoc.io/static/org.assertj/assertj-core/3.22.0/org/assertj/core/api/AbstractIterableAssert.html#satisfiesExactlyForProxy(java.util.function.Consumer%5B%5D)

**Comment by lukas-krecan** [comment]
Cool, thanks a lot for the analysis. It will take me some time to walk through it.

**Comment by lukas-krecan** [comment]
Fixed in 2.32.0 I did not get into marking `satisfiesXXX` as deprecated but it makes sense to do it. Thanks again

**Comment by glhez** [comment]
Hello,

I realize one of my example does not compile:

```
assertThatJson(json).inPath("$.[0]").isArray().elements(0).isString().startsWith("a");
------------------------------------------------^^^^^^
```

The "elements" should have been element(0).

Also, while enabling back my test, the element(0) fails to work due to the JsonAssert evaluating the node as JSON. This correspond to my comment https://github.com/lukas-krecan/JsonUnit/issues/474#issuecomment-1039442937 

However, I tested against my fork (which I deleted since).

I do remember that I've added a parameter to JsonAssert to tell the assert that it was already a json node:

```
assertThatJson(json).inPath("$.[0]")
                                .isArray()
                                .element(0)
                                .isString() // fails because String is evaluated as JSON
                                .startsWith("a");
```
